nomad/event_sink_manager.go

```diff
@@ -402,148 +379,134 @@ func (m *ManagedSink) Run() error {
 					return
 				}
 
-				// check if we should reload or just reset
-
-				reload := m.ResetWatchSet()
-
-				if !reload {
-					continue
-				}
-				// Cancel the subscription scoped context
-				m.cancelFn()
-
-				// wait until the reset was done
-				select {
-				case <-m.stopCtx.Done():
-					return
-				case <-m.doneReset:
-				case <-exitCh:
-				}
+				// WatchCh was triggered, reset the the WatchCh and potentially
+				// cancel the currentSinkCtx to reload changes
+				m.resetSink()
 			}
 		}
 	}()
 
-LOOP:
+START:
+	// Subscribe to event broker and establish SinkWriter
+	err := m.subscribe()
+	if err != nil {
+		return err
+	}
+
+	currentSinkCtx := m.sinkCtx
 	for {
-		events, err := m.Subscription.Next(m.sinkCtx)
+		events, err := m.subscription.Next(currentSinkCtx)
 		if err != nil {
 			// Shutting down, exit gracefully
 			if m.stopCtx.Err() != nil {
 				return m.stopCtx.Err()
 			}
 
-			// Reloadable error, reload and restart
-			if err == stream.ErrSubscriptionClosed || err == context.Canceled {
-				if err := m.Reload(); err != nil {
-					return err
-				}
-				goto LOOP
-			}
-			return err
+			m.logger.Info("received error from managed sink subscription, reloading sink", "error", err)
+			goto START
 		}
 
-		err = m.SinkWriter.Send(m.sinkCtx, &events)
+		// Send the events to the writer with stopCtx to cancel if manager is
+		// shutting down
+		err = m.sinkWriter.Send(m.stopCtx, &events)
 		if err != nil {
 			if strings.Contains(err.Error(), context.Canceled.Error()) {
+				// if the context is canceled continue and let the subscription
+				// error checking handle if we should reload or exit
 				continue
 			}
-			m.l.Warn("Failed to send event to sink", "sink", m.Sink.ID, "error", err)
+			m.logger.Warn("Failed to send event to sink", "sink", m.Sink.ID, "error", err)
 			continue
+		} else {
+			// Update the last successful index sent
+			atomic.StoreUint64(&m.lastSuccess, events.Index)
 		}
-		// Update the last successful index sent
-		atomic.StoreUint64(&m.lastSuccess, events.Index)
 	}
 }
 
-// ResetWatchSet resets the managed sinks watchCh after a change has been made
+// resetSink resets the managed sinks watchCh after a change has been made
 // to the event sink. It returns whether or not the subscription and sink
 // should be reloaded.
-func (m *ManagedSink) ResetWatchSet() bool {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+func (m *ManagedSink) resetSink() {
+	currentSubCancel := m.sinkCancelFn
 
+	// Fetch the Event Sink
 	ws := m.stateFn().NewWatchSet()
 	sink, err := m.stateFn().EventSinkByID(ws, m.Sink.ID)
 	if err != nil {
-		return true
+		// Log the error, continue to set the sink to nil
+		m.logger.Error("error querying for latest event sink", "sink", m.Sink.ID, "error", err)
 	}
 
-	if sink == nil {
-		return true
-	}
-
-	// Update the watchCh
-	m.watchCh = ws.WatchCh(m.sinkCtx)
+	// Set our Sink to the new one
+	oldSink := m.Sink
+	m.Sink = sink
+	m.watchCh = ws.WatchCh(m.stopCtx)
```

Comment:
There's a small ordering change we made here; should this `watchCh` only get changed if the sink changed? In the previous version of this the `watchCh` only got changed if `m.Sink != nil`, whereas now the nil check comes afterwards.